liquibase-standard/src/main/java/liquibase/snapshot/SnapshotGeneratorChain.java

```diff
@@ -27,14 +31,17 @@ public SnapshotGeneratorChain(SortedSet<SnapshotGenerator> snapshotGenerators) {
     }
 
     /**
-     * This calls all the non-replaced {@link SnapshotGenerator} in the chain, by comparison order
-     * Only the first generator in the chain is allowed to create a new instance of T
-     * Subsequent generators must modify the instance or call the chain if the provided object is not handled,
-     * otherwise a {@link DatabaseException} is thrown
+     * This calls all the non-replaced {@link SnapshotGenerator} in the chain, by comparison order.
+     * <p>
+     * During the chain processing, the snapshot method returns an object that will be passed to the next generator in the chain.
+     * The object can be the same object that was passed to the generator, or a new object, given that:
+     * - the generator copies the attributes of the provided object to a new object and returns the new object (liquibase-hibernate extension does this in https://github.com/liquibase/liquibase-hibernate/blob/main/src/main/java/liquibase/ext/hibernate/snapshot/IndexSnapshotGenerator.java#L31 )
+     * - the generator returns the existing instance (they can e.g. set new attributes to it)
+     * <p>
+     * Snapshot generators that do not abide by the previous rules must be the first to be called for a given object type and there are not more than one of these per object type.
+     * Note that this can get tricky as extensions can bring their own snapshot generators to the mix breaking core generators.
```

Comment:
Should we add a way for a generator to tell us what kind of generator it is? One which plays by the rules, or one which doesn't? Then if we see two which say they don't play by the rules, we throw an exception or warning?

Reply:
@filipelautert all the existing problematic generators have been / are about to be fixed, right?

It would be nice to add some default validation to snapshot generators, so that "bad" generators cannot be introduced anymore. Maybe the validation logic could be as simple as:

1. create a snapshot instance (of type `T`, where `T extends DatabaseObject`)
2. add some custom attributes (`snapshot.setAttribute("some-random-key", "thing")`)
3. call the generator `snapshot` method
4. check that the result still contains the custom attribute?

Is that too simplistic?
I'm not sure where that validation should take place though. 
And also: if a generator fails to pass the validation, does the execution just stop there and report the violation? Does it ignore the generator?